pkg/serial/handshake_test.go

```diff
@@ -0,0 +1,387 @@
+	select {
+	case <-f.deadline:
+		return 0, errors.New("Timeout error")
+	default:
+	}
+	count := 0
+	for count < len(b) {
+		select {
+		case v := <-f.c:
+			b[count] = v
+			count++
+		case _, ok := <-f.deadline:
+			if ok {
+				close(f.deadline)
+			}
+			return 0, errors.New("Timeout error")
+		default:
+			if count == 0 {
+				time.Sleep(time.Millisecond)
+			} else {
+				return count, nil
+			}
+		}
+	}
+	return count, nil
+}
+
+type BiChannel struct {
+	L *BlockingSendReceiver
+	R *BlockingSendReceiver
+}
+
+func (bc *BiChannel) Write(b []byte) (int, error) {
+	return bc.L.Send(b)
+}
+
+func (bc *BiChannel) Read(b []byte) (int, error) {
+	return bc.R.Receive(b)
+}
+
+func NewFakeConnection(t time.Duration) (*BiChannel, *BiChannel) {
+	l := NewBlockingSendReceiver().Timeout(t)
+	r := NewBlockingSendReceiver().Timeout(t)
+	return &BiChannel{L: l, R: r}, &BiChannel{L: r, R: l}
+}
+
+func TestHandshakeServerNormalCaseScenario(t *testing.T) {
+	log.SetLevel(log.InfoLevel)
+	clientConn, serverConn := NewFakeConnection(time.Second)
+
+	go func() {
+		buf := make([]byte, 10)
+		clientConn.Write([]byte{flagSyn, 200})
+
+		if n, e := clientConn.Read(buf); e != nil || n != 3 {
+			t.Errorf("Only 3 bytes are expected: %x, received: %d", buf[:n], n)
+			return
+		}
+
+		if buf[0] != flagAck || buf[1] != 201 || buf[2] < 128 {
+			t.Errorf("Error, unexpected data: %x", buf[:3])
+			return
+		}
+		clientConn.Write([]byte{flagAck, incrementByte(buf[2])})
+	}()
+
+	if e := HandshakeServer(serverConn); e != nil {
+		t.Errorf("Unexpected error: %v", e)
+	}
+}
+
+func TestHandshakeServerLotsOfTrashOnTheLine(t *testing.T) {
+	log.SetLevel(log.InfoLevel)
+	clientConn, serverConn := NewFakeConnection(time.Second * 3)
+
+	go func() {
+		buf := make([]byte, 10)
+
+		// Do not send too many bytes, otherwise "write" will block on server side due too many flagNak.
+		x := "sdfkgn sdflkjsfdfdgis dfgs"
+		for i := 0; i < 10; i++ {
+			x += x
+		}
+		clientConn.Write([]byte(x))
+		clientConn.Write([]byte{flagSyn, 200})
+
+		for {
+			n, e := clientConn.Read(buf)
+			if e != nil {
+				t.Errorf("Unexpected server error: %v", e)
+				return
+			}
+
+			if n < 3 {
+				continue
+			}
+
+			data := buf[n-3:]
+			if data[0] == flagNak {
+				continue
+			}
+
+			if data[0] != flagAck || data[1] != 201 || data[2] < 128 {
+				t.Errorf("Error, unexpected data: %x", data[:3])
+				return
+			}
+			clientConn.Write([]byte{flagAck, incrementByte(data[2])})
+		}
+	}()
+
+	for {
+		if e := HandshakeServer(serverConn); e == nil {
+			break
+		} else {
+			if _, ok := e.(*HandshakeError); !ok {
+				t.Errorf("Unexpected error: %v", e)
+				return
+			}
+		}
+	}
+}
+
+func TestHandshakeServerComportSync(t *testing.T) {
+	log.SetLevel(log.InfoLevel)
+	clientConn, serverConn := NewFakeConnection(time.Second * 3)
+
+	go func() {
+		buf := make([]byte, 10)
+		clientConn.Write([]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22})
```

Comment:
Can you add a test that seeds the line (so the client will read them) with these bytes, and has the client and server hooked together as in `TestHandshakeServerVsClient` - that's the normal case that we see between containerVM and applianceVM. It's the containerVM boot sequence that generates this byte sequence.